Added skin tone support for Emojis

mananjadhav opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@stitesExpensify Task complete. Can you review, please?

### Details
Extended the existing Emoji Picker to add Skin Tone support.

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
$ #4323 

### Tests
1. Tested several emojis with all skin tones
2. Verified NVP is stored for Preferred Skin tone
3. Verified NVP is available on refresh on the same device and other devices
4. Tested Filter + Skintone
5. Verified missing skin tones for all the emojis
6. Tested default and null handling for emoji skin tones in case skin tone not available
7. UI elements in terms of selection, on hover, refreshing the list tested
8. UI matched with the mockup shared
9. Tested with Landscape mode for Mobile Devices

<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### QA Steps
1. Go to chats and open EmojiPicker
2. With the first attempt the default skin tone should show up.
3. Choose the skin tone and verify that all the Emojis with Skin tone support have the chosen skin tone
4. Select any emoji and it should render with the chosen skin tone
5. Change the skin tone and the emojis should have the updated skin tone
6. Login with another device and ensure that the chosen Emoji skin tone is reflected by default
7. Try deleting the emoji and see that it doesn't require two deletes. 
8. Attempt to load the Emoji picker in the Mobile landscape view as well and see that emoji picker doesn't break
9. Search the emojis with keywords. These should also show up the emojis in chosen skin tone only

<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
https://user-images.githubusercontent.com/3069065/128575992-db9d5d94-10fd-42f8-8c48-c7903ea8657f.mp4

#### Mobile Web
https://user-images.githubusercontent.com/3069065/128576083-eddcdd13-7add-4c5a-9682-a309225fd048.mp4

<img width="748" alt="emoji-skintone-mweb-landscape" src="https://user-images.githubusercontent.com/3069065/128577164-e2131ece-76e1-4a67-9c58-0042998be35f.png">

**Updated**
<img width="722" alt="Screenshot 2021-09-09 at 9 19 51 AM" src="https://user-images.githubusercontent.com/3069065/132620632-d675ca8f-e25d-49a3-bd73-7afee8fce9e9.png">


#### Desktop
https://user-images.githubusercontent.com/3069065/128576109-651981a0-7d1b-463f-b168-1e1f263df712.mp4

#### iOS
https://user-images.githubusercontent.com/3069065/128576156-2787a270-bdec-4fad-9db2-c4240b5c5211.mp4
<img width="764" alt="emoji-skintone-ios-landscape" src="https://user-images.githubusercontent.com/3069065/128577479-26270f45-afd9-40e6-8b2f-cede0012d25b.png">

**Updated**
<img width="757" alt="Screenshot 2021-09-09 at 9 26 41 AM" src="https://user-images.githubusercontent.com/3069065/132620620-bc132b79-ea8b-4982-abf1-b531d427b1ae.png">


#### Android
https://user-images.githubusercontent.com/3069065/128576868-6800164f-5b8a-49a5-8495-793eff6a839c.mp4

![emoji-skintone-android-landscape](https://user-images.githubusercontent.com/3069065/128577119-8221f11b-3465-4ddd-9ae1-08cc034cddde.jpeg)

**Updated**
<img width="1105" alt="Screenshot 2021-09-09 at 9 32 11 AM" src="https://user-images.githubusercontent.com/3069065/132620607-7888cf4d-d3de-4bce-bd18-b52ba7040b80.png">

mananjadhav committed:
fix(emoji-skintone): Added skintones list to emoji

mananjadhav committed:
fix(emoji-skintone): Added key in Onyx for preferred skintone

mananjadhav committed:
fix(emoji-skintone): Props and state setup for emojis

mananjadhav committed:
fix(emoji-skintone): Added view with skintone list and updated emoji code

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into fix/emoji-skintone

mananjadhav committed:
fix(emoji-skintone): Added default hand skintone

mananjadhav committed:
refactor(emoji-skintone): Updated skintone array format

mananjadhav committed:
feat(emoji-skintone): Moved skintone picker to a new render function

mananjadhav committed:
chore(emoji-skintone): Updated translations

mananjadhav committed:
fix(emoj-skintone): Added helpertext with style changes

mananjadhav committed:
fix(emoji-skintone): Updated item style for skintone picker

mananjadhav committed:
fix(emoji-skintone): Fixed index check for type selection

mananjadhav committed:
fix(emoji-skintone): Rerender emoji on code change

mananjadhav committed:
fix(emoji-skintone): Store skinTone preference on the server side

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

mananjadhav committed:
fix(emoji-skintone): Moved preferredSkinTone from state to props

mananjadhav committed:
fix(emoji-skintone): Moved the api calls and default key for nvp

mananjadhav committed:
fix(emoji-skintone): null handling for skintone

mananjadhav committed:
fix(emoji-skintone): 0 handling for skintone

mananjadhav committed:
docs(emoji-skintone): Updated comments and jsdocs

mananjadhav committed:
refact(emoji-skintone): Moved getSkinToneEmoji to new file

mananjadhav committed:
fix(emoji-skintone): Added skinPicker to native

mananjadhav committed:
refactor(emoji-skintone): Moved SkinTone picker to a new component

mananjadhav committed:
fix(emoji-skintone): Fixed errors with EmojiSkinTone picker

mananjadhav committed:
refact(emoji-skintone): replaced function by new skin tone list component

mananjadhav committed:
fix(emoji-skintone): Fixed highlight logic for skin tone list

mananjadhav committed:
refactor(emoji-skintone): Removed unused references and code cleanup

mananjadhav committed:
refactor(emoji-skintone): Replaced skin tone list func with component

mananjadhav committed:
fix(emoji-skintone): Updated func binding

mananjadhav committed:
refactor(emoji-skintone): wrapped nvp call with User.setPreferredSkinTone action

mananjadhav committed:
refactor(emoji-skintone): Replaced state.preferredSkinTone to props.preferredSkinTone

mananjadhav commented:
@stitesExpensify @puneetlath I've completed the task. Please review the changes. 

Need help with two things:
1. Need to verify the design and send feedback if any deviation
2. Translations for Spanish. Have added using Google translate for the time being.

stitesExpensify commented on src/pages/home/report/EmojiSkinToneList.js:
```diff
@@ -0,0 +1,108 @@
+const propTypes = {
+
+    /** Visibility flag for skin tone picker */
+    isSkinToneListVisible: PropTypes.bool,
```
Why is this controlled by props instead of just being controlled with state locally in this component?

mananjadhav replied:
Because the trigger will always be from the main picker component. 

Still, want to maintain it in the state?

stitesExpensify replied:
Why is the trigger in the main picker component though? Can't we just put all of that logic in here?

mananjadhav replied:
Don't remember why I did it that way. Moved to state.

stitesExpensify commented on src/pages/home/report/EmojiSkinToneList.js:
```diff
@@ -0,0 +1,108 @@
+                      [styles.emojiHeaderStyle, styles.emojiExtraSmall, styles.textAlignCenter, styles.pv0]
+                      }
+                    >
+                        {props.translate('emojiPicker.setPreferredSkinTone')}
```
I don't think we need this, it's pretty self explanatory IMO

mananjadhav replied:
Removed

stitesExpensify reviewed (changes requested):
The functionality seems to work really well! Just some small-ish comments

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

mananjadhav committed:
fix(emoji-skintone): Removed unwanted labels

mananjadhav committed:
fix(emoji-skintone): line rewritten for better readability

mananjadhav committed:
refactor(emoji-skintone): Removed extra checks and changed fetch for emojiindex function

mananjadhav committed:
refactor(emoji-skintone): Changed 'default' handling

mananjadhav commented on src/languages/en.js:
```diff
@@ -609,4 +609,8 @@ export default {
     },
+    emojiPicker: {
+        skinTonePickerLabel: 'Choose skin tone',
+        setPreferredSkinTone: 'Set your preferred default skin tone',
```
Removed.

mananjadhav commented on src/libs/actions/User.js:
```diff
@@ -69,7 +70,7 @@ function getBetas() {
     API.Get({
         returnValueList: 'account, loginList, nameValuePairs',
-        nvpNames: CONST.NVP.PAYPAL_ME_ADDRESS,
+        nvpNames: `${CONST.NVP.PAYPAL_ME_ADDRESS}, ${CONST.NVP.PREFERRED_SKIN_TONE}`,
```
Done.

mananjadhav commented on src/languages/es.js:
```diff
@@ -611,4 +611,8 @@ export default {
     },
+    emojiPicker: {
+        skinTonePickerLabel: 'Elige el tono de piel',
+        setPreferredSkinTone: 'Establezca su tono de piel predeterminado preferido',
```
Removed

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/index.native.js:
```diff
@@ -36,12 +47,29 @@ class EmojiPickerMenu extends Component {
+     */
+    setPreferredSkinTone(skinTone) {
+        this.setState({showSkinToneList: false});
+        if (this.props.updatePreferredSkinTone) {
```
Removed.

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -95,6 +107,17 @@ class EmojiPickerMenu extends Component {
+     */
+    setPreferredSkinTone(skinTone) {
+        this.setState({showSkinToneList: false});
+        if (this.props.updatePreferredSkinTone) {
```
Removed.

mananjadhav committed:
refactor(emoji-skintone): Updated emojiCode check

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -320,11 +343,16 @@ class EmojiPickerMenu extends Component {
         }
 
+        let emojiCode = code;
+        if (types && typeof this.props.preferredSkinTone === 'number' && types[this.props.preferredSkinTone]) {
```
After checking again, we don't need `default` check too. Updated with ternary check.

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/getSkinToneEmojiFromIndex.js:
```diff
@@ -0,0 +1,16 @@
+ * @param {Number} skinToneIndex
+ * @returns {String}
+ */
+const getSkinToneEmojiFromIndex = (skinToneIndex) => {
```
Worked. Updated accordingly.

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/getSkinToneEmojiFromIndex.js:
```diff
@@ -0,0 +1,16 @@
+ * @returns {String}
+ */
+const getSkinToneEmojiFromIndex = (skinToneIndex) => {
+    if (typeof skinToneIndex !== 'number') {
```
NA. Changed function body

mananjadhav committed:
refactor(emoji-skintone): Change setPreference logic

mananjadhav commented on src/libs/actions/User.js:
```diff
@@ -84,6 +85,10 @@ function getUserDetails() {
+
+            const preferredSkinTone = lodashGet(response, `nameValuePairs.${CONST.NVP.PREFERRED_SKIN_TONE}`, {});
+            Onyx.merge(ONYXKEYS.NVP_PREFERRED_SKIN_TONE,
+                typeof preferredSkinTone === 'number' ? preferredSkinTone : 'default');
```
Instead, I've updated this line with our function call.

```
Onyx.merge(ONYXKEYS.NVP_PREFERRED_SKIN_TONE,
                getSkinToneEmojiFromIndex(preferredSkinTone).skinTone);
```

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

# Conflicts:
#	src/styles/utilities/spacing.js

mananjadhav commented:
@stitesExpensify PR updated, with 1-2 comments. Please have a look.

shawnborton commented:
Adding Michelle here as a reviewer so she can provide design/UX feedback. 

I think it would be nice if the emoji modal didn't jump in height when you were interacting with the skin tone emojis, so ideally the transition between states keeps the modal at the same height.

mananjadhav commented:
@shawnborton are you suggesting that the skinTone list should be collapsible or be an overlay on top of the existing modal? And not be an additional section below ?

shawnborton commented:
I'm suggesting that the small area below the list of emojis (let's call it the "tone picker") has a fixed height, so no matter the state of the "tone picker" the area maintains the same height. Actually I wonder if making a quick Figma prototype might be helpful here?

mananjadhav commented:
Yeah, a Figma would help. 

and honestly, we should soon add the categories support also the way we've shown in the mockup.

Apart from these, there are few other design aspects we should consider (probably as a separate scope):
1. Increase the size of the Emoji and reduce the spacing (maybe have 6 in a row)
2. Size of the emoji on the chat after entering is also too small afaict
3. Show the emoji keyword on hover
4. Landscape mode on mobile devices, emoji picker takes the whole screen. Hence, no place to hide the picker without entering the emoji
5. Adding an `X` (close) symbol to clear the search text instantly
6. Show frequently used Emojis (You'll see users use less than 10% of the emoji universe)
7. Bonus: Add a custom Emoji with Expensify logo

What do you think @michelle-thompson

shawnborton commented:
Those are good suggestions, but I would consider most of them out of scope for this particular task.

mananjadhav commented:
I agree. These would also need heavy involvement from Design.

puneetlath commented:
@mananjadhav can you create an issue for each suggestion you have for emoji picker improvements so that we can discuss each of them individually?

michelle-thompson commented:
Do we need to write `Set your preferred default skin tone`? We already say `Choose skin tone` before that, so I think it's obvious what to do. 

Perhaps we do something like the below?
![image](https://user-images.githubusercontent.com/43361905/128897084-84f089e4-b4cc-4eac-8a82-cb3eec296ac3.png)

puneetlath commented:
What if we had the text say: 
>Update default skin tone

That way it's clear that the hand you're seeing is your current default and whatever you choose will now be your default across emojis. 

We can also discuss this in #expensify-open-source if we want to have a more realtime discussion about it.

michelle-thompson commented:
I like that suggestion. Super minor, but I'd prefer `Choose your default skin tone` because it sounds a bit more casual.

puneetlath commented:
Ah ok. I liked `update` because it implies that a default is already set is already set. Whereas `choose` sounds like an action you need to do. But I don't feel super strongly. Good with whatever you decide is best.

michelle-thompson commented:
I don't feel super strongly here either, so curious if @shawnborton has a preference.

shawnborton commented:
Throwing a third option into the mix - "Change default skin tone" - I think it feels more casual than "Update" and it addresses Puneet's comment here:

> That way it's clear that the hand you're seeing is your current default and whatever you choose will now be your default across emojis.

michelle-thompson commented:
Works for me 👍

stitesExpensify commented:
> Do we need to write Set your preferred default skin tone? We already say Choose skin tone before that, so I think it's obvious what to do.

I already had him remove the secondary text :) 

> Change default skin tone

I like that. @mananjadhav I'll get a translation for that

mananjadhav committed:
fix(emoji-skintone): Change skintone picker label

mananjadhav committed:
fix(emoji-skintone): Moved skinTonevisibility to state

mananjadhav commented:
Hi Team,
I've updated the label and it is a fixed height modal now. Let me know if any other changes.

https://user-images.githubusercontent.com/3069065/128969024-4981551a-d8e6-430b-8e4a-a71fbe52192f.mp4

shawnborton commented:
Looks pretty good to me, nice work! Curious for others' thoughts too.

shawnborton commented:
I like 6. For 7, I think that just warrants a larger discussion. At some point I think we'd like to support this (uploading custom emojis), but I am not sure if we want to put the effort into that now. Probably a good thing to bring to #expensify-open-source in Slack. 

That being said, I think a good habit for you to get into is to bring all of these quick proposals to that room to get a quick gut check from the team on whether or not it's worth pursuing. Thanks!

michelle-thompson commented:
This looks great, thank you @mananjadhav! Agreed that 6 is something that would good to implement now.

mananjadhav commented:
@shawnborton  @michelle-thompson  Thanks for the feedback. I'll be posting pt6 in a while.

@shawnborton I don't have access to Expensify slack channel I guess.

shawnborton commented:
@mananjadhav can you send an email to `contributors@expensify.com` with a subject of "Request to join Slack channel" and a message of something like:

```
Hi,

Can you please invite me to the Slack Channel?

Thanks!
```

mananjadhav commented:
@shawnborton That easy? Done. XD

stitesExpensify commented on src/languages/es.js:
```diff
@@ -611,4 +611,7 @@ export default {
         growlMessageNoPersonalPolicy: 'No he podido encontrar una póliza personal con la que asociar esta llamada a las Guías, compruebe su conexión e inténtelo de nuevo.',
     },
+    emojiPicker: {
+        skinTonePickerLabel: 'Cambiar el tono de piel predeterminado',
```
Elige el tono de piel por defecto

stitesExpensify commented on src/pages/home/report/EmojiSkinToneList.js:
```diff
@@ -25,28 +22,31 @@ const propTypes = {
 
 const EmojiSkinToneList = (props) => {
     const selectedEmoji = getSkinToneEmojiFromIndex(props.preferredSkinTone);
     const [highlightedIndex, setHighlightedIndex] = useState(selectedEmoji.skinTone);
```
We don't use hooks like `useState` in our codebase, so this component will have to be changed. Sorry for not noticing it in my earlier review 😞

mananjadhav replied:
Ohh okay. I'll change it to a class component. Any other changes?

stitesExpensify replied:
Not that I can see at the moment. Thanks!

mananjadhav committed:
refactor(emoji-skintone): Change spanish translation

mananjadhav committed:
fix(emoji-skintone): Changed pickerlist from function to class component

mananjadhav commented:
PR updated

stitesExpensify commented on src/pages/home/report/EmojiSkinToneList.js:
```diff
@@ -26,76 +27,88 @@ const propTypes = {
 EmojiSkinToneList.propTypes = propTypes;
-EmojiSkinToneList.defaultProps = defaultProps;
-
+EmojiSkinToneList.defaultProps = {};
```
We can just get rid of this line right? Since all props are required?

mananjadhav replied:
Updated

mananjadhav committed:
refactor(emoji-skintone): Removed defaultProps

stitesExpensify commented:
Hey @mananjadhav PR looks good, just so you know though, i'm not going to be able to test and merge until next week though due to other priorities.

mananjadhav commented:
@stitesExpensify Cool. Meanwhile, do you want me to run any specific tests at my end?

stitesExpensify commented:
Found an interesting bug where trying to delete a modified emoji makes it go back to yellow on first press, and then gets deleted on the second press.

https://user-images.githubusercontent.com/42391420/129761417-91ad42b7-8065-4043-8b45-92e095ea6961.mp4

mananjadhav commented:
That is a weird one 😕  I'll check this

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

mananjadhav commented:
@stitesExpensify I figured mostly what the issue is. An emoji with skin tone is nothing but the actual emoji + skin tone Unicode. Now with the backspace event, first it deletes the skin tone Unicode, and then with the second one it deletes the default Emoji Unicode. 

I am thinking of overriding the backspace event with `triggerHotkeyActions` and appropriately delete the whole emoji. What do you think? Any better suggestions?

mananjadhav commented:
@stitesExpensify Found one more thing related to this.

We're sending the emojis as `emoji={`${emojiCode}\uFE0F`}`. Can you tell me why is `\uFE0F` added? I am sure there must be a reason. But if I remove the Unicode the emoji still work and the delete also works perfectly fine.


cc: @rdjuric  You've worked on Emoji earlier. Can you shed some light?

mananjadhav commented:
@stitesExpensify Did you get a chance to look at my last comment?

stitesExpensify commented:
Hi @mananjadhav I don't know why it was added. Does the git blame say anything in the commit log?

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

mananjadhav committed:
fix(emoji-skintone): Handled delete bug with skintone emojis

mananjadhav commented:
I can see @rdjuric's message:

```
We add '\uFE0F' to our Unicode to force the correct emoji presentation (VS16) on emojis that also have a text-style presentation (VS15).
```

Afaik, skin tone emojis don't have a text-style presentation. I've added a fix for this. Can @stitesExpensify and @rdjuric confirm if [this commit](https://github.com/Expensify/App/pull/4456/commits/4c9c9227792abf98ef7f0167edbd07bf297c0e35) works fine?

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

mananjadhav committed:
fix(emoji-skintone): Fixed unwanted unicodes causing delete issues

mananjadhav committed:
fix(emoji-skintone): Updated text emojis to unicode emoji styles

mananjadhav committed:
fix(emoji-skintone): Removed unwanted unicode from skintone picker

mananjadhav committed:
fix(emoji-skintone): Removed comment

mananjadhav commented:
@stitesExpensify Sorry for bombarding the thread with messages. I've finally fixed the issue for all the emojis. It seems we added only text-style codes for certain emojis. I've updated the emoji list and removed `\uFE0F` from the code.

This should also fix #4623 where delete is causing an issue for flag(2 codepoint emojis).

puneetlath commented:
@mananjadhav looks like you have some merge conflicts.

mananjadhav commented:
@puneetlath Sure. I'll resolve those by tomorrow.

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

# Conflicts:
#	src/ONYXKEYS.js
#	src/pages/home/report/EmojiPickerMenu/index.js
#	src/pages/home/report/EmojiPickerMenu/index.native.js
#	src/pages/home/report/EmojiPickerMenuItem.js

mananjadhav committed:
fix(emoji-skintone): Updated skintone alignment based on new changes

mananjadhav commented:
With the conflicts, we had some alignment issues. I've resolved the conflict and fixed the alignment issue as well.

@puneetlath @stitesExpensify Anything else pending on this one?


<img width="1212" alt="Screenshot 2021-09-03 at 3 40 41 AM" src="https://user-images.githubusercontent.com/3069065/131923157-625a587e-0afb-4b5f-b43f-29cdbceb45b6.png">
<img width="364" alt="Screenshot 2021-09-03 at 3 42 49 AM" src="https://user-images.githubusercontent.com/3069065/131923208-283744db-ba25-4475-bd6c-269fd7810379.png">
<img width="367" alt="Screenshot 2021-09-03 at 3 42 41 AM" src="https://user-images.githubusercontent.com/3069065/131923219-85820342-2889-4a58-8797-3fd8e0ad2373.png">

stitesExpensify commented:
Hi @mananjadhav I'll review this tonight or tomorrow morning. Sorry for the delays

stitesExpensify reviewed (commented):
Code looks good and tests good! Just a couple of small changes and we'll be good to go

mananjadhav commented on assets/emojis.js:
```diff
@@ -6173,7 +6193,7 @@ const emojis = [
         ],
     },
     {
-        code: '☄',
```
Changed plaintext emojis to emoji Unicode format. Looks a bit tricky to visualize but if you check https://emojipedia.org/. Paste both of them you'll see the difference.

stitesExpensify replied:
Gotcha. What is the purpose of that?

mananjadhav replied:
To resolve the delete issue. Primarily just to support these limited emojis, we were unnecessarily adding `f0ef` to all the emojis, causing the delete to require two backspaces. This solves the linked delete flag issue as well.

mananjadhav committed:
fix(emoji-skintone): Matched propnames

mananjadhav committed:
refactor(emoji-skintone): Changed preferred_skin_tone key for Onyx and NVP

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -369,11 +382,17 @@ class EmojiPickerMenu extends Component {
                         />
                     )}
+                <EmojiSkinToneList
+                    setPreferredSkinTone={this.props.updatePreferredSkinTone}
```
I guess I had a conflict earlier and then I refactored but I didn't change this. Updated now.

mananjadhav commented on src/ONYXKEYS.js:
```diff
@@ -120,6 +120,9 @@ export default {
     REIMBURSEMENT_ACCOUNT: 'reimbursementAccount',
 
+    // Store preferred skintone for emoji
+    NVP_PREFERRED_SKIN_TONE: 'preferredSkinTone',
```
Updated

mananjadhav commented on src/CONST.js:
```diff
@@ -227,6 +227,7 @@ const CONST = {
         FAILED_BANK_ACCOUNT_VALIDATIONS_PREFIX: 'private_failedBankValidations_',
         BANK_ACCOUNT_GET_THROTTLED: 'private_throttledHistory_BankAccount_Get',
         PREFERRED_LOCALE: 'preferredLocale',
+        PREFERRED_SKIN_TONE: 'expensify_preferredSkinTone',
```
Updated

mananjadhav commented on src/pages/home/report/EmojiPickerMenu/index.js:
```diff
@@ -21,6 +22,12 @@ const propTypes = {
     forwardedRef: PropTypes.func,
 
+    /** Stores user's preferred skin tone */
+    preferredSkinTone: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
```
Yeah, we had `undefined` earlier but not Onyx call ensures we get `default`. Made required.

mananjadhav commented:
@stitesExpensify clarifying again:

1.  `CONST.NVP. PREFERRED_EMOJI_SKIN_TONE: 'expensify_preferredEmojiSkinTone'`
2.  `ONXYKEYS.PREFERRED_EMOJI_SKIN_TONE: 'preferredSkinTone',`

mananjadhav committed:
refactor(emoji-skintone): Update keynames for onyx and nvp

stitesExpensify commented:
Let's just do 
```
CONST.NVP. PREFERRED_EMOJI_SKIN_TONE: 'expensify_preferredEmojiSkinTone'
ONXYKEYS.PREFERRED_EMOJI_SKIN_TONE: 'preferredEmojiSkinTone',
```

mananjadhav commented:
Ohh my, Sorry for the confusion but that's what I've done in the PR. I made a typo while putting it in a comment. Can you please review and close it?

stitesExpensify commented:
Everything is looking great! One more thing I found while testing is that the selector is not the same height as the button which makes the picker jump around

https://user-images.githubusercontent.com/42391420/132562721-9a3fe348-862a-4ed4-a85b-23131f753148.mp4

mananjadhav commented:
Okay noted. I'll check and update it.

stitesExpensify commented:
Thanks! I think after that we'll get this merged!

mananjadhav committed:
fix(emoji-skintone): Added separate styling for Emoji Skin tone title

mananjadhav committed:
Merge branch 'main' of https://github.com/mananjadhav/App into feat/emoji-skintone

# Conflicts:
#	src/libs/actions/User.js

mananjadhav commented:
@stitesExpensify PR updated.


https://user-images.githubusercontent.com/3069065/132604583-c8c83a44-6db6-440a-8c47-9d6de398ee00.mov

mananjadhav committed:
fix(emoji-skintone): Handle landscape view for mobile web

mananjadhav committed:
fix(emoji-skintone): Removed the modal changes for landscape view

mananjadhav committed:
fix(emoji-skintone): Added landscape view check for mobile app

mananjadhav commented:
@stitesExpensify I found another issue that the Skintone picker wasn't visible in the mobile landscape mode. This was because of another fix in the PR where we added some top margin so that the landscape mode picker works fine. I've added a fix for that too. I guess now it is good to be merged with all the testing :)


<img width="722" alt="Screenshot 2021-09-09 at 9 19 51 AM" src="https://user-images.githubusercontent.com/3069065/132620354-0847e11b-2632-4ef9-bbf6-adb1464fee35.png">

<img width="757" alt="Screenshot 2021-09-09 at 9 26 41 AM" src="https://user-images.githubusercontent.com/3069065/132620341-6301fda2-f70c-48d2-97b8-53a2afee5965.png">

<img width="1105" alt="Screenshot 2021-09-09 at 9 32 11 AM" src="https://user-images.githubusercontent.com/3069065/132620578-fcc92e35-b59f-42b2-b73f-c3db213445f3.png">

stitesExpensify reviewed (approved):
LGTM!

stitesExpensify commented:
Can you write some QA steps? After that I'll merge :)

mananjadhav commented:
@stitesExpensify Updated QA steps. Anything I've missed out?

stitesExpensify commented:
Nope those look great. Time to ship it! :smile:

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1218882715) to staging by @stitesExpensify in version: 1.0.95-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

botify commented:
This has been deployed to production and is now subject to a 7-day regression period.
If no regressions arise, payment will be issued on 2021-09-17.  :confetti_ball:

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1222459337) to production by @roryabraham in version: 1.0.96-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mananjadhav commented on src/pages/home/report/EmojiSkinToneList.js:
```diff
@@ -0,0 +1,109 @@
+                            }
+                            style={[
+                                styles.pv1,
+                                styles.flex1,
```
Thanks for pointing out.
